### DISCUSSION TITLE
A working usecase that avoids deadlocks when using gevent and SQLAlchemy together

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7925] [Usage Questions]
TL;DR my last comment in this thread describes a usecase/pattern of access that works for me.

I'm using postgres and gevent.

I've decided I'm utterly confused by autobegin, autocommit, recommended session lifetimes, and the what the "right" usage pattern is with SQLA for anything but trivial usecases. My starting point here is https://github.com/sqlalchemy/sqlalchemy/discussions/6921 and the hole I'm trying to dig myself out of involves a randomly hung app which, AFAIK, is due to being stuck in idle transactions I never asked to be in.

As I understand it, in order for me to avoid having multiple copies of the same database item in my app, that item must be in a session. Yet, the pattern suggested in #6921 is to do a bunch of database operations and then close the session they're performed in.

Meanwhile, objects are related to each other, so merely traversing a (possibly lazily loaded) relationship may trigger a database operation. And, as far as I can tell, such traversals may trigger a begin, thus ending up in an idle transaction. So what do I do? If I have a complex object graph and need to visit parts of it randomly, doesn't that mean I must keep a long lived session open, and have all my objects in it, so that as I move around the object graph, it can fetch the parts of the graph I'm visiting? If I do that, doesn't it mean I'm going to be randomly stuck in open transactions that may deadlock other operations?

What's the right pattern for that kind of app, where I want the object with a given id, whether fetched from the DB or looked up from a cache to always be the same object and where I do not want SQLA leaving idle transactions open?

**Comment by CaselIT** [comment]
Hi,

By default the dbapi spec requires that any operation on the db is in a transaction. That's not something sqlalchemy invented.
Some drivers allow an autocommit mode that's not in the dbapi spec.
See https://docs.sqlalchemy.org/en/14/core/connections.html#dbapi-autocommit to enable it.

This should let you have a long lived session that's not holding on a idle transaction

**Reply by aathan** [reply]
I guess by implication you're confirming the "right" pattern if I don't want duplicate objects all over the place is to keep them in a long lived session?

I am aware of the autocommit API but there's a lot of obfuscating content out there regarding postgresql in particular relative to autobegin and autocommit, what works, and what doesn't (including the fairly old post referenced from #6921 : https://www.oddbird.net/2014/06/14/sqlalchemy-postgres-autocommit/ ).

By the way, something changed between 1.4.31 or 1.4.32 and 1.4.35 ... the code under .31 and .32 fairly consistently throws a None doesn't have a "twophase" attribute, whereas .35 doesn't. I guess something about the operation ordering is exercising a race differently.

**Reply by aathan** [reply]
I'm trying all permutations of versions and committing and not committing and autocommit not autocommit etc to try and get a handle on what is affecting these behaviors. This is the kind of error message that strikes fear into the heart of the uninitiated:

```SAWarning: Attribute history events accumulated on 2 previously clean instances within inner-flush event handlers have been reset, and will not result in database updates. Consider using set_committed_value() within inner-flush event handlers to avoid this warning.```

I'm doing inner-flushes? What's an inner-flush? What does this error mean exactly? I definitely don't have inner-flush event handlers, so does this have something to do with changing an object's values from another greenlet, while it's in a commit()ed in a second greenlet?

**Reply by CaselIT** [reply]
> I guess by implication you're confirming the \"right\" pattern if I don't want duplicate objects all over the place is to keep them in a long lived session?

The usual pattern, as mike suggest, is to use short lived sessions. I don't know what you are doing so I can't suggest a best way of doing it.

I was just suggesting autocommit if you wanted to keep a long lived session anyway.

In any case you seem to be hitting wired edge cases that no one else has raised, so I would really suggest looking at the tutorial and go from there.

**Reply by aathan** [reply]
1) I was of course aware of the autocommit API but your message gave me the push I needed to look at that again. The solution in my case was to go deeper into the app's framework to find where the sessions are created and break the requirement that all sessions come from the same factory -- I then set only *one* of the sessions (the long lived one) into autocommit.

2) What I'd like to fully understand (and maybe documented if it's not already) is what the pattern is for the use case I described, irrespective of what's "usual." That usecase being:   I have some (large) set of objects and relationships between them stored in the DB, there is random access into that set, and random traversals across the relationships, and there should be only one copy in memory per identity in the DB.

Is the right way to achieve that a long lived Session in which all long lived in-memory objects are queried from, and/or who's relationships are traversed while they are owned by that session -- and into which all new objects are add()ed after being persisted?

I feel like that's a yes or no question, but maybe there are nuances I'm not aware of?

If the answer is "yes" then it may be worth documenting that in the presence of threading or co-routines, due to autobegin, this pattern can lead to database deadlocks from idle transactions that remain open following the automated activity of SQLA upon object graph traversals, and one way to solve that issue is to set the one long lived session into autocommit mode, and do all your writes from other sessions? This of course would require that no implicit fetches/traversals to related objects occur during the write (when the object is in fact in the writer session).  Instead of ensuring those accesses are limited, writes could be done by *copying* the object-graph-to-be-written into a writer session thereby leaving the "real" objects in the reader session.

I'm not sure if there is a utility method that could help you traverse all dirty objects related to the object you want to write. That's what I would use to do "copy then write." I guess such a method would similar to an expunge() but where the object remains in the session, and instead you get a copy() callback or somesuch.

The code I have now instead "expunge()s" the objects that should be written from the long-lived query session, add()s them to a writer session, writes them, and then expunge()s them from the writer session and puts them back into the reader session.

3) Like the 100 flushes error message, perhaps that SAWarning: I posted above could use a clarifying re-wording. I'm guessing it is (like that 100 flush message) assuming what's happening is due to "inner-flush event handlers" (which I have none) but in fact the condition it's describing may occur irrespective of "inner-flush event handlers."  Maybe what it's telling me is "object instances being flushed changed value mid-flush after already being marked clean" and one way that could happen is if their values change after one of those "100 flush"es that the session does to bring itself to clean state?  In other words, by "inner-flush" what's being referred to is one of those 100 flushes??? One way this could happen without "inner-flush event handlers" is if a co-routine is doing the value changes.

**Reply by zzzeek** [reply]
> 
>     2. What I'd like to fully understand (and maybe documented if it's not already) is what the pattern is for the use case I described, irrespective of what's "usual." That usecase being:   I have some (large) set of objects and relationships between them stored in the DB, there is random access into that set, and random traversals across the relationships, and there should be only one copy in memory per identity in the DB.

if your application uses concurrency like gevent or threading, this is not possible.   Global mutable state does not mix with threading/greenlet concurrency, full stop.    if you want state that is global to all threads etc., you need to store it in a form that is immutable at the level of individual objects, like a cache or a fixed JSON-like structure.  ORM Session will not work.


> 
> 
> Is the right way to achieve that a long lived Session in which all long lived in-memory objects are queried from, and/or who's relationships are traversed while they are owned by that session -- and into which all new objects are add()ed after being persisted?

it seems like you want some pattern like an object cache.  the pattern for this is that objects are never mutated outside of a session.  when you want to work with objects in the context of a transaction, you create your Session (database transaction), merge in the objects you need using session.merge() (makes a session-local copy of the object), then work with those session-local copies.   The example suite at https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.dogpile_caching shows this pattern in the context of a cache backend, but the objects could just as well be in some local collection, as long as the global objects are never mutated in any way and never attached directly to a Session as long as their object identity is globally shared.

> 
> If the answer is "yes" then it may be worth documenting that in the presence of threading or co-routines, due to autobegin, this pattern can lead to database deadlocks from idle transactions that remain open following the automated activity of SQLA upon object graph traversals, and one way to solve that issue is to set the one long lived session into autocommit mode, and do all your writes from other sessions?

autocommit mode is not a solution to this problem and is also a long-deprecated pattern in any case.    The documentation tries to make clear at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#is-the-session-thread-safe how the Session should be considered in a concurrent context.

> 
>     3. Like the 100 flushes error message, perhaps that SAWarning: I posted above could use a clarifying re-wording. I'm guessing it is (like that 100 flush message) assuming what's happening is due to "inner-flush event handlers" (which I have none) but in fact the condition it's describing may occur irrespective of "inner-flush event handlers."  Maybe what it's telling me is "object instances being flushed changed value mid-flush after already being marked clean" and one way that could happen is if their values change after one of those "100 flush"es that the session does to bring itself to clean state?  In other words, by "inner-flush" what's being referred to is one of those 100 flushes??? One way this could happen without "inner-flush event handlers" is if a co-routine is doing the value changes.


if you are sharing the state of a single Session between concurrent threads / gevent greenlets, behavior is undefined.   Catastrophic conditions like this should be expected.

**Comment by aathan** [comment]
Before I respond, would you clarify whether "Global mutable state does not mix with threading/greenlet concurrency, full stop" refers to SQLA things or is it meant as a "computability" statement about concurrent programming in general?

**Reply by zzzeek** [reply]
both

**Reply by aathan** [reply]
Fine. I'm hoping to get at a clear description of a usecase by discussing the issues in small bites. So, I'm just going to ignore that statement so we can progress the conversation until it runs into that statement head on.

* I read the documents you linked to both just now, and had previously read them. Maybe not everything stuck.
* For level setting purposes, I’m an experienced programmer and have maintained large complex concurrent systems written in a variety of lanuages.
* Below, I offer statements that reflect my understanding with the expectation that you'll refute the incorrect ones.

Having said that, I wonder if we're mixing issues of concurrency "inside" a Session/SQLA with the more general issues of concurrency in the app, outside of Sessions/SQLA.

We should unmix those.

There is a fundamental "problem" in the general concurrency area that arises from the shared database. An idle read lock may block other apps, let alone other greenlets in the same app, whether they too are using SQLA, the same or different connections to the DB, or have one or many threads etc.

The idle read lock deadlocking issue is distinct from SQLA's programming model. I see it arise frequently. However, it involves SQLA only in that auto-begin semantics in the presence of Pythonic syntactic sugar makes the read locks potentially invisible. A program that fetches Object o through Session s may access x.some_attr and it cannot be known at the time that line of code is written, that this will definitively NOT have begun a transaction. However, it is impractical to assume every such line has done so.

Do we agree so far?

**Reply by CaselIT** [reply]
I'm not sure where you want to arrive..

> I'm hoping to get at a clear description of a usecase by discussing the issues in small bites

This just seem to make everyone waste time. Please provide an example (MCVE) of what you are doing and what issues you are having.

**Reply by zzzeek** [reply]
> Having said that, I wonder if we're mixing issues of concurrency "inside" a Session/SQLA with the more general issues of concurrency in the app, outside of Sessions/SQLA.
> 
> We should unmix those.

I'm only offering information here.  SQLAlchemy Session is not threadsafe/gevent-concurrency safe.  If you are doing that, and you are getting failure modes with confusing error messages that don't seem to make sense, those are not bugs in SQLAlchemy.  So that's that part of it.


> 
> There is a fundamental "problem" in the general concurrency area that arises from the shared database. An idle read lock may block other apps, let alone other greenlets in the same app, whether they too are using SQLA, the same or different connections to the DB, or have one or many threads etc.

The database, assuming you aren't using SQLite, has an sophisticated concurrency-control system as part of its architecture.   So yes if you are building a comprehensive concurrency-control system in your gevent app, sure, global mutable state is "fine".   However the ORM Session itself is never going to be a good fit as the source of that global mutable state since it refers to a single database transaction at a time.


> 
> The idle read lock deadlocking issue is distinct from SQLA's programming model. I see it arise frequently. However, it involves SQLA only in that auto-begin semantics in the presence of Pythonic syntactic sugar makes the read locks potentially invisible. A program that fetches Object o through Session s may access x.some_attr and it cannot be known at the time that line of code is written, that this will definitively NOT have begun a transaction. However, it is impractical to assume every such line has done so.

I am not really following this but it seems like you're saying that "lazyloads can start a transaction", which is true from a database perspective.  But lazyloads can't start a Session().  If you don't have a Session(), there's no lazyloads, and there's no transactions starting on the fly.    Hence the recommended pattern is that the Session itself is considered to be "a transaction, in progress", so as long as a particular Session exists, that Session should be short lived and used in only one thread/greenlet at a time.

The above paragraph is the main rule to be following.   

> Do we agree so far?

I agree with @CaselIT that it's not clear what your goals are in this discussion, if this is to get information about how the Session works, which I think we've provided, or you're unsatisfied with how the Session works, for which we are not really in the position to be changing things and there are lots of other ORMs to choose from that don't go anywhere near our often-controversial usage pattern, or you're just trying to get more ideas on how to restructure your application, for which the guidance would be move the Session outside of the role of global mutable state and that's as much as we can offer on this end.

**Reply by aathan** [reply]
Fair enough. I was just trying to reduce the degrees of freedom in each response.

Anyway, @zzzeek 's response is probably sufficient so if y'all don't find the conversation useful then I won't continue it past this response. As always, thanks for your time. @zzzeek commented somewhere that these issues frequently come up, so I hope this conversation has at least helped someone.

https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html is an important document. Anyone running into this tread should  read that.

A concrete suggestion: There's lots of documentation relevant to these issues, but maybe more references and links could be added from certain places (like lazy_load APIs etc). Concurrency issues in general could be discussed more frequently in the docs now that  asyncio and gevent are mainstream. E.g., https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html and the custom load rules section etc could mention concurrency usage patterns.  Once I become more confident about all of it, and what you've blessed, I'll try to give back to the community by submitting some doc PRs.

Anyway, here is my SQLA usage pattern in my concurrent app. It allows for one copy in memory of each item in the DB, and using lazy-load relationships. Maybe it won't receive your blessing, but it works, and it argues for not removing autocommit.

* Set your sessions to expire_on_commit=False
* Carefully configure relationships and their loading (e.g. lazyload)
* Serialize ORM DB fetches (queries) onto a single long lived Session set to autocommit (call this your query_session)
* Be aware that editing an object while it is owned by the query_session may cause those edits to be autocommit'ed 
* To write to the DB, expunge changed objects from the query_session and add them to a new autobegin write_session. Then, commit() the write_session and expunge() the objects from the write_session (or close() it) and add() the objects back to the query_session
* Serialize your use of write_session(s) and/or queries via the query_session so that objects and their relations are never add()ed to multiple sessions (or you'll get an exception)

Given this is in a concurrent app, there must be some access pattern that manages concurrent changes/access to objects. That pattern is outside the scope of the above bullet points but my version of it motivates them. In it, either my threads are directly reading attributes out of the very same object another thread may have changed and that's ok, or those changes are protected in critical sections, or new objects not yet know to other threads are being created so those new objects are safe. I gather changed objects into a collection, put that collection onto a work queue, and have a single writer thread perform the expunge dance I described above. The autocommit on the read_session ensures that "implicit" database operations based on natural Pythonic attribute accesses won't accidentally leave an idle read lock open.

This pattern works well for me, irrespective of whether autocommit is considered a deprecated pattern. Autocommit seems extremely useful to making lazyload and expired attribute access workable in a concurrent SQLA program. It means that code can access attributes/relationships without worrying that an idle read transaction is left open on the underlying database. It is not possible to know at the time you write a line that accesses an ORM attribute, whether at runtime, the attribute access will result in a database operation. That's because this issue is controlled by configuration of the rest of the ecosystem, including the expire_on_commit() settings of any session that ORM object has passed through at runtime.
